src/workloads/sharding/MultiShardTransactions.yml

```diff
@@ -0,0 +1,343 @@
+          OperationCommand:
+            movePrimary: *Database1
+            to: rs1
+
+- Name: LoadInitialData
+  Type: MonotonicLoader
+  Threads: 4
+  Phases:
+  - &Nop {Nop: true}
+  - Repeat: 1
+    Threads: 4
+    BatchSize: 1000
+    DocumentCount: *DocumentCount
+    CollectionCount: 2
+    Database: *Database0
+    Document: &Document
+      counter: 0
+      padding: {^FastRandomString: {length: {^RandomInt: {min: *ApproxDocumentSize50Pct, max: *ApproxDocumentSize}}}}
+  - Repeat: 1
+    Threads: 4
+    BatchSize: 1000
+    DocumentCount: *DocumentCount
+    CollectionCount: 2
+    Database: *Database1
+    Document: *Document
+  - *Nop
+  - *Nop
+  - *Nop
+  - *Nop
+
+- Name: Transactions
+  Type: CrudActor
+  Threads: 16
+  Database: *Database0
+  Phases:
+    - *Nop
+    - *Nop
+    - *Nop
+    - MetricsName: ShardedCollectionUpdate
+      Duration: 1 minute
+      Collection: Collection0
+      ThrowOnFailure: false  # Transactions are allowed to fail.
+      RecordFailure: true  # We still want to record the metrics even if the transaction is rolled back.
+      Operations:
+      - OperationName: withTransaction
+        OperationCommand:
+          OperationsInTransaction:
+          - OperationName: updateOne
+            OperationCommand:
+              Filter: &RandomIdFilter {_id: {^RandomInt: {min: 1, max: *DocumentCount}}}
+              Update: {$inc: {counter: 1}}
+          - OperationName: updateOne
+            OperationCommand:
+              Filter: *RandomIdFilter
+              Update: {$inc: {counter: 1}}
+          - OperationName: updateOne
+            OperationCommand:
+              Filter: *RandomIdFilter
+              Update: {$inc: {counter: 1}}
+          - OperationName: updateOne
+            OperationCommand:
+              Filter: *RandomIdFilter
+              Update: {$inc: {counter: 1}}
+          - OperationName: updateOne
+            OperationCommand:
+              Filter: *RandomIdFilter
+              Update: {$inc: {counter: 1}}
+          - OperationName: updateOne
+            OperationCommand:
+              Filter: *RandomIdFilter
+              Update: {$inc: {counter: 1}}
+          - OperationName: updateOne
+            OperationCommand:
+              Filter: *RandomIdFilter
+              Update: {$inc: {counter: 1}}
+          - OperationName: updateOne
+            OperationCommand:
+              Filter: *RandomIdFilter
+              Update: {$inc: {counter: 1}}
+          - OperationName: updateOne
+            OperationCommand:
+              Filter: *RandomIdFilter
+              Update: {$inc: {counter: 1}}
+          - OperationName: updateOne
+            OperationCommand:
+              Filter: *RandomIdFilter
+              Update: {$inc: {counter: 1}}
+          - OperationName: updateOne
+            OperationCommand:
+              Filter: *RandomIdFilter
+              Update: {$inc: {counter: 1}}
+    - MetricsName: ShardedAndNonShardedCollectionsUpdate
```

Comment:
Taken from the ticket description:
> Tests should include:
> 1. Single database transaction that targets two non-primary shards of a sharded collection.
> 2. Single database transaction that targets a non-primary shard of a sharded collection and a non-sharded collection on primary shard.
> 3. Multi-database transaction that targets two non-sharded collections on different primary shards.
> 4. Multi-database transaction that targets two non-primary shards of two sharded collections collections.

[question] Could you explain why we want to explicitly test the performance of a transaction that targets non-primary shards? What's the significance of not being the primary shard when it comes to change-stream?

Assuming that we still want to test the 4 categories above, I believe the current version only covers the third category correctly (under `DifferentDatabasesNonShardedCollectionsUpdate`). The reason is that with hashed sharding, an updater against a random collection can end up targeting any shard that owns data for the collection (i.e. primary or non-primary shard).

Reply:
I have read the design and the thread that you had with Andy, Max and Rachita in the sharding channel.  I see how you want to make the perf workload run transactions that target only non-primary shards, specifically because the current design intends to make the coordinator additionally send a writeEndOfTransaction command to the primary shard so you want to measure the performance impact of engaging another shard. I spoke to Max about this. We both think it isn't very important to make the transactions not have the primary shard as a participant since we just need to measure is the extra cost of going to the primary shard to write down the list of namespaces. Given this, we probably should just make the workload run any cross-shard transactions, i.e. not care about the participants (keep the current hashed sharding setup).